File: src/eqn/ast/EqnAstNodeCustomFunction.java
package eqn.ast;

import eqn.parser.exception.EqnException;
import org.jetbrains.annotations.NotNull;

import java.util.Vector;

public class EqnAstNodeCustomFunction extends EqnAstNode {

    public int arity;
    Vector<EqnAstNode> operands;

    public EqnAstNodeCustomFunction(String value) {
        super(value, Type.CustomFunction);
        operands = new Vector<>(0);
    }


    public EqnAstNodeCustomFunction(String value, int arity) {
        super(value, Type.CustomFunction);
        operands = new Vector<>(0);
        this.arity = arity;
    }

    public void addOperand(@NotNull EqnAstNode operand) {
        this.operands.add(operand);
    }

    @Override
    public String toString() {
        StringBuffer stringBuffer = new StringBuffer(value + '(');
        for (int i = 0; i < operands.size() - 1; ++i) {
            stringBuffer.append(operands.elementAt(i).toString() + ", ");
        }
        stringBuffer.append(operands.lastElement().toString());

        stringBuffer.append(")");
        return new String(stringBuffer);
    }

    @Override
    public double evaluate() throws EqnException {
        throw new EqnException("Custom function evaluate not yet implemented (make this class abstract)!");
    }

    @Override
    public EqnAstNode simplify() throws EqnException {
        for (int i = 0; i != arity; ++i) {
            operands.set(i, operands.elementAt(i).simplify());
        }
        return this;
    }
}

File: src/eqn/ast/EqnAstNodeCustomFunction.kt
package eqn.ast

import eqn.ast.base.EqnAstNode
import eqn.ast.base.EqnAstNodeArbitraryArity
import eqn.parser.exception.EqnException

class EqnAstNodeCustomFunction(value: String) : EqnAstNodeArbitraryArity(value, Type.CustomFunction, PrecedenceType.Function) {
    @Throws(EqnException::class)
    override fun evaluate(arguments: Map<String, Double>?): Double {
        throw EqnException("Custom function evaluate not yet implemented (make this class abstract)!")
    }

    @Throws(EqnException::class)
    override fun simplify(arguments: Map<String, Double>?): EqnAstNode {
        for (i in 0 until arity()) {
            operands[i] = operands.elementAt(i).simplify(arguments)
        }
        return this
    }
}